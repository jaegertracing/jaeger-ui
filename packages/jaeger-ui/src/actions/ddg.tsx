// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import _identity from 'lodash/identity';
import { createActions, ActionFunctionAny, Action } from 'redux-actions';

import {
  TDdgAddViewModifierPayload,
  TDdgClearViewModifiersFromIndicesPayload,
  TDdgRemoveViewModifierFromIndicesPayload,
  TDdgRemoveViewModifierPayload,
  TDdgViewModifierRemovalPayload,
} from '../model/ddg/types';
import generateActionTypes from '../utils/generate-action-types';

export const actionTypes = generateActionTypes('@jaeger-ui/DEEP-DEPENDENCY-GRAPH', [
  'ADD_VIEW_MODIFIER',
  'CLEAR_VIEW_MODIFIERS_FROM_INDICES',
  'REMOVE_VIEW_MODIFIER',
  'REMOVE_VIEW_MODIFIER_FROM_INDICES',
]);

const addViewModifier: (kwarg: TDdgAddViewModifierPayload) => TDdgAddViewModifierPayload = _identity;
const clearViewModifiersFromIndices: (
  kwarg: TDdgClearViewModifiersFromIndicesPayload
) => TDdgClearViewModifiersFromIndicesPayload = _identity;
const removeViewModifier: (kwarg: TDdgRemoveViewModifierPayload) => TDdgRemoveViewModifierPayload = _identity;
const removeViewModifierFromIndices: (
  kwarg: TDdgRemoveViewModifierFromIndicesPayload
) => TDdgRemoveViewModifierFromIndicesPayload = _identity;

const fullActions = createActions<TDdgAddViewModifierPayload | TDdgViewModifierRemovalPayload>({
  [actionTypes.ADD_VIEW_MODIFIER]: addViewModifier,
  [actionTypes.CLEAR_VIEW_MODIFIERS_FROM_INDICES]: clearViewModifiersFromIndices,
  [actionTypes.REMOVE_VIEW_MODIFIER]: removeViewModifier,
  [actionTypes.REMOVE_VIEW_MODIFIER_FROM_INDICES]: removeViewModifierFromIndices,
});

type FullActionsType = {
  jaegerUi: {
    deepDependencyGraph: Record<string, ActionFunctionAny<Action<TDdgViewModifierRemovalPayload>>>;
  };
};

export default (fullActions as unknown as FullActionsType).jaegerUi.deepDependencyGraph;
