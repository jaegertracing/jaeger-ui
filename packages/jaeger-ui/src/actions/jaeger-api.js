// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createAction } from 'redux-actions';
import JaegerAPI from '../api/jaeger';

export const fetchTrace = createAction(
  '@JAEGER_API/FETCH_TRACE',
  id => JaegerAPI.fetchTrace(id),
  id => ({ id })
);

export const fetchMultipleTraces = createAction(
  '@JAEGER_API/FETCH_MULTIPLE_TRACES',
  ids => JaegerAPI.searchTraces({ traceID: ids }),
  ids => ({ ids })
);

export const archiveTrace = createAction(
  '@JAEGER_API/ARCHIVE_TRACE',
  id => JaegerAPI.archiveTrace(id),
  id => ({ id })
);

export const searchTraces = createAction(
  '@JAEGER_API/SEARCH_TRACES',
  query => JaegerAPI.searchTraces(query),
  query => ({ query })
);

export const fetchServices = createAction('@JAEGER_API/FETCH_SERVICES', () => JaegerAPI.fetchServices());

export const fetchServiceOperations = createAction(
  '@JAEGER_API/FETCH_SERVICE_OPERATIONS',
  serviceName => JaegerAPI.fetchServiceOperations(serviceName),
  serviceName => ({ serviceName })
);

export const fetchDeepDependencyGraph = createAction(
  '@JAEGER_API/FETCH_DEEP_DEPENDENCY_GRAPH',
  query => JaegerAPI.fetchDeepDependencyGraph(query),
  query => ({ query })
);

export const fetchDependencies = createAction('@JAEGER_API/FETCH_DEPENDENCIES', () =>
  JaegerAPI.fetchDependencies()
);
