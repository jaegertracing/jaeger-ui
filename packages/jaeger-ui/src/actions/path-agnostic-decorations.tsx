// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import _get from 'lodash/get';
import _memoize from 'lodash/memoize';
import _set from 'lodash/set';
import memoize from 'lru-memoize';
import { createActions, ActionFunctionAny, Action } from 'redux-actions';

import JaegerAPI from '../api/jaeger';
import { TNewData, TPathAgnosticDecorationSchema } from '../model/path-agnostic-decorations/types';
import { getConfigValue } from '../utils/config/get-config';
import generateActionTypes from '../utils/generate-action-types';
import stringSupplant from '../utils/stringSupplant';

// wrapping JaegerAPI.fetchDecoration is necessary for tests to properly mock inside memoization
const fetchDecoration = memoize(10)((url: string) => JaegerAPI.fetchDecoration(url));

export const actionTypes = generateActionTypes('@jaeger-ui/PATH_AGNOSTIC_DECORATIONS', ['GET_DECORATION']);

export const getDecorationSchema = _memoize((id: string): TPathAgnosticDecorationSchema | undefined => {
  const schemas = getConfigValue('pathAgnosticDecorations') as TPathAgnosticDecorationSchema[] | undefined;
  if (!schemas) return undefined;
  return schemas.find(s => s.id === id);
});

let doneCount: undefined | number;
let pendingCount: undefined | number;
let pendingData: undefined | TNewData;
let pendingPromise: undefined | Promise<TNewData>;
let resolve: undefined | ((arg: TNewData) => void);

// Bespoke memoization-adjacent solution necessary as this should return `undefined`, not an old promise, on
// duplicate calls
// exported for tests
export const _processed = new Map<string, Map<string, Set<string | undefined>>>();

export function getDecoration(
  id: string,
  service: string,
  operation?: string
): Promise<TNewData> | undefined {
  const processedID = _processed.get(id);
  if (!processedID) {
    _processed.set(id, new Map<string, Set<string | undefined>>([[service, new Set([operation])]]));
  } else {
    const processedService = processedID.get(service);
    if (!processedService) processedID.set(service, new Set([operation]));
    else if (!processedService.has(operation)) processedService.add(operation);
    else return undefined;
  }

  const schema = getDecorationSchema(id);
  if (!schema) return undefined;

  const returnPromise = !resolve || !pendingPromise;
  if (returnPromise) {
    pendingPromise = new Promise<TNewData>(res => {
      resolve = res;
    });
  }

  pendingCount = pendingCount ? pendingCount + 1 : 1;
  const { summaryUrl, opSummaryUrl, summaryPath, opSummaryPath } = schema;
  let promise: Promise<Record<string, any>>;
  let getPath: string;
  let setPath: string;
  if (opSummaryPath && opSummaryUrl && operation) {
    promise = fetchDecoration(stringSupplant(opSummaryUrl, { service, operation }));
    getPath = stringSupplant(opSummaryPath, { service, operation });
    setPath = `${id}.withOp.${service}.${operation}`;
  } else {
    promise = fetchDecoration(stringSupplant(summaryUrl, { service }));
    getPath = stringSupplant(summaryPath, { service });
    setPath = `${id}.withoutOp.${service}`;
  }

  promise
    .then(res => {
      return _get(res, getPath, `\`${getPath}\` not found in response`);
    })
    .catch(err => {
      return `Unable to fetch decoration: ${err.message || err}`;
    })
    .then(value => {
      if (!pendingData) pendingData = {};
      _set(pendingData, setPath, value);
      doneCount = doneCount ? doneCount + 1 : 1;

      if (doneCount === pendingCount) {
        if (resolve) resolve(pendingData);
        // istanbul ignore next : Unreachable error to appease TS, resolve made to exist at top at function
        else throw new Error('`resolve` unexpectedly undefined');

        // eslint-disable-next-line no-multi-assign
        doneCount = pendingCount = pendingData = pendingPromise = resolve = undefined;
      }
    });

  if (returnPromise) return pendingPromise;
  return undefined;
}

const fullActions = createActions<Promise<TNewData> | undefined>({
  [actionTypes.GET_DECORATION]: getDecoration,
});

export default (fullActions as any).jaegerUi.pathAgnosticDecorations as Record<
  string,
  ActionFunctionAny<Action<Promise<TNewData> | undefined>>
>;
