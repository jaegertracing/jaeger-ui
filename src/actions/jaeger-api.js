// Copyright (c) 2017 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import { createAction } from 'redux-actions';
import JaegerAPI from '../api/jaeger';

/**
 * async wrapper to get the api object in case we're in demo mode.
 */

export const fetchConfig = createAction('@JAEGER_API/FETCH_CONFIG', () => JaegerAPI.fetchConfig());

export const fetchTrace = createAction(
  '@JAEGER_API/FETCH_TRACE',
  id => JaegerAPI.fetchTrace(id),
  id => ({ id })
);

export const searchTraces = createAction(
  '@JAEGER_API/SEARCH_TRACES',
  query => JaegerAPI.searchTraces(query),
  query => ({ query })
);

export const fetchServices = createAction('@JAEGER_API/FETCH_SERVICES', () => JaegerAPI.fetchServices());

export const fetchServiceOperations = createAction(
  '@JAEGER_API/FETCH_SERVICE_OPERATIONS',
  serviceName => JaegerAPI.fetchServiceOperations(serviceName),
  serviceName => ({ serviceName })
);

export const fetchDependencies = createAction('@JAEGER_API/FETCH_DEPENDENCIES', () =>
  JaegerAPI.fetchDependencies()
);
