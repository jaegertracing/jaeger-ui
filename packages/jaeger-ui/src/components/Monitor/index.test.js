// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import store from 'store';
import MonitorATMPage from '.';
import servicesReducer from '../../reducers/services';
import metricsReducer from '../../reducers/metrics';
import * as jaegerApiActions from '../../actions/jaeger-api';

// --- Mock Modules ---
// Mock the actions module
jest.mock('../../actions/jaeger-api');
// Mock the 'store' npm package
jest.mock('store');

// --- Mock References ---
// Reference the mocked actions module
const mockedJaegerApiActions = jaegerApiActions;
const mockedStorage = store;

// --- Redux Setup ---
const rootReducer = combineReducers({
  services: servicesReducer,
  metrics: metricsReducer,
});
const initialState = {
  services: { services: [], loading: false, error: null, operations: {} },
  metrics: {
    loading: false,
    operationMetricsLoading: false,
    isATMActivated: true, // Prevent rendering EmptyState initially
    opsError: { opsCalls: null, opsErrors: null, opsLatencies: null },
    serviceOpsMetrics: [],
    serviceMetrics: { service_latencies: null, service_error_rate: null, service_call_rate: null },
    serviceError: {
      service_latencies_50: null,
      service_latencies_75: null,
      service_latencies_95: null,
      service_error_rate: null,
      service_call_rate: null,
    },
  },
};
const mockStore = createStore(rootReducer, initialState);

describe('<MonitorATMPage>', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    // --- Configure Mocks ---
    // Configure mock implementations on the *actions* module
    // Use mockImplementation or mockReturnValue as appropriate for actions
    mockedJaegerApiActions.fetchServices.mockImplementation(() => ({ type: 'FETCH_SERVICES_MOCK' }));
    mockedJaegerApiActions.fetchAllServiceMetrics.mockImplementation(() => ({
      type: 'FETCH_ALL_METRICS_MOCK',
    }));
    mockedJaegerApiActions.fetchAggregatedServiceMetrics.mockImplementation(() => ({
      type: 'FETCH_AGG_METRICS_MOCK',
    }));

    // Configure store mocks
    mockedStorage.get.mockImplementation(key => {
      if (key === 'lastAtmSearchService') return '';
      if (key === 'lastAtmSearchSpanKind') return 'server';
      if (key === 'lastAtmSearchTimeframe') return 3600000;
      return null;
    });
    mockedStorage.set.mockImplementation(() => {});
  });

  it('does not explode and renders initial elements', () => {
    const { container } = render(
      <Provider store={mockStore}>
        <MemoryRouter>
          <MonitorATMPage />
        </MemoryRouter>
      </Provider>
    );

    expect(container).toBeDefined();

    // Check calls on the mocked *actions*
    expect(mockedJaegerApiActions.fetchServices).toHaveBeenCalledTimes(1);
    // Metrics fetches should NOT happen on initial mount because initialState.services is empty.
    expect(mockedJaegerApiActions.fetchAllServiceMetrics).not.toHaveBeenCalled();
    expect(mockedJaegerApiActions.fetchAggregatedServiceMetrics).not.toHaveBeenCalled();

    // Check main headings and selectors
    expect(screen.getByRole('heading', { name: /Service/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Span Kind/i })).toBeInTheDocument();
    expect(screen.getAllByRole('combobox').length).toBeGreaterThanOrEqual(3);

    // Check graph headers using roles
    expect(screen.getByRole('heading', { name: /Latency/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Error rate/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Request rate/i })).toBeInTheDocument();

    // Check table section
    expect(screen.getByRole('heading', { name: /Operations metrics/i })).toBeInTheDocument();
    expect(screen.getByRole('table')).toBeInTheDocument();
  });

  it('renders EmptyState when isATMActivated is false', () => {
    // Create a specific state for this test where ATM is not activated
    const emptyStateInitialState = {
      ...initialState, // Spread the common initial state
      metrics: {
        ...initialState.metrics, // Spread the common metrics state
        isATMActivated: false, // Override to false
      },
    };
    const emptyStateStore = createStore(rootReducer, emptyStateInitialState);

    // Render with the modified store
    render(
      <Provider store={emptyStateStore}>
        <MemoryRouter>
          <MonitorATMPage />
        </MemoryRouter>
      </Provider>
    );

    // Assert EmptyState component is rendered (e.g., by checking for its unique elements)
    // Using the image alt text as a likely unique identifier
    expect(screen.getByAltText('jaeger-monitor-tab-preview')).toBeInTheDocument();

    // Assert main UI elements are NOT rendered
    // Use exact match for 'Service' heading to avoid matching EmptyState content
    expect(screen.queryByRole('heading', { name: /^Service$/i })).not.toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: /Latency/i })).not.toBeInTheDocument();
    expect(screen.queryByRole('table')).not.toBeInTheDocument();

    // fetchServices is called unconditionally in componentDidMount
    expect(mockedJaegerApiActions.fetchServices).toHaveBeenCalledTimes(1);
    // Metrics fetches should NOT happen if services array is initially empty
    expect(mockedJaegerApiActions.fetchAllServiceMetrics).not.toHaveBeenCalled();
    expect(mockedJaegerApiActions.fetchAggregatedServiceMetrics).not.toHaveBeenCalled();
  });
});
