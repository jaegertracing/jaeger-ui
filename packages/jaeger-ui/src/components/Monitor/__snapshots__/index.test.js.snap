// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonitorATMPage> Component is displayed 1`] = `<Connect(ReduxForm) />`;
